jerry-main/fuzzilli.c

```diff
@@ -0,0 +1,371 @@
+    if (__shmem == MAP_FAILED)
+    {
+      fprintf(stderr, "Failed to mmap shared memory region\n");
+      _exit(-1);
+    }
+  }
+
+  __sanitizer_cov_reset_edgeguards();
+
+  __shmem->num_edges = (uint32_t)(stop - start);
+  printf("[COV] edge counters initialized. Shared memory: %s with %u edges\n", shm_key, __shmem->num_edges);
+}
+
+void __sanitizer_cov_trace_pc_guard(uint32_t *guard)
+{
+  // There's a small race condition here: if this function executes in two threads for the same
+  // edge at the same time, the first thread might disable the edge (by setting the guard to zero)
+  // before the second thread fetches the guard value (and thus the index). However, our
+  // instrumentation ignores the first edge (see libcoverage.c) and so the race is unproblematic.
+  uint32_t index = *guard;
+  __shmem->edges[index / 8] |= 1 << (index % 8);
+  *guard = 0;
+}
+
+// We have to assume that the fuzzer will be able to call this function e.g. by
+// enumerating the properties of the global object and eval'ing them. As such
+// this function is implemented in a way that requires passing some magic value
+// as first argument (with the idea being that the fuzzer won't be able to
+// generate this value) which then also acts as a selector for the operation
+// to perform.
+jerry_value_t
+jerryx_handler_fuzzilli(const jerry_value_t func_obj_val, /**< function object */
+                        const jerry_value_t this_p,       /**< this arg */
+                        const jerry_value_t args_p[],     /**< function arguments */
+                        const jerry_length_t args_cnt)    /**< number of function arguments */
+{
+  (void)func_obj_val; /* unused */
+  (void)this_p;       /* unused */
+
+  if (args_cnt < 1 || !jerry_value_is_string(args_p[0]))
+  {
+    return jerry_create_error(JERRY_ERROR_TYPE, (jerry_char_t *)"Expected a string");
+  }
+
+  jerry_value_t ret_val = jerry_create_undefined ();
+  jerry_size_t str_size = jerry_get_utf8_string_size (args_p[0]);
+  jerry_char_t *operation = jerry_heap_alloc (str_size * sizeof(jerry_char_t));
+
+  if (operation == NULL || jerry_string_to_utf8_char_buffer(args_p[0], operation, str_size) != str_size)
+  {
+    return jerry_create_error(JERRY_ERROR_RANGE, (jerry_char_t *)"Internal error");
+  }
+
+  if (strcmp((char *)operation, "FUZZILLI_CRASH") == 0)
+  {
+    if (args_cnt == 2 && jerry_value_is_number(args_p[1]))
+    {
+      int arg = (int)jerry_get_number_value(args_p[1]);
+      switch (arg)
+      {
+      case 0:
+        *((int *)0x41414141) = 0x1337;
+        break;
+      default:
+        DCHECK(false);
+        break;
+      }
+    }
+  }
+  else if (strcmp((char *)operation, "FUZZILLI_PRINT") == 0)
+  {
+    static FILE *fzliout;
+    fzliout = fdopen(REPRL_DWFD, "w");
+    if (!fzliout)
+    {
+      fprintf(stderr, "Fuzzer output channel not available, printing to stdout instead\n");
+      fzliout = stdout;
+    }
+
+    /* Based on the jerryx_handler_print handler */
```

Comment:
Is this a verbatim copy of `jerryx_handler_print`? (Disclaimer: I haven't checked.) If it is, wouldn't it be better to redirect stdout temporarily to `REPRL_DWFD`, then call `jerryx_handler_print`, and finally, reset stdout file descriptor to its original file? Something using `dup`/`dup2`. (Disclaimer: I haven't tried.) It would look like a bit of magic, but it would prevent copy-paste-induced future maintenance burden.

Reply:
I was also thinking about it, but `dup2` wasn't working. JerryScript internally uses `putchar`, which uses `FILE *stdout` stream. Instead of `dup()`-ing the descriptor I made a backup of `stdout`, replaced it by the stream representing `REPRL_DWFD`, called the original `jerryx_handler_print ` handler and restored the `stdout`.